libvast/src/detail/posix.cpp

```diff
@@ -171,6 +171,33 @@ int uds_recv_fd(int socket) {
 
 VAST_DIAGNOSTIC_POP
 
+int uds_sendmsg(int socket, const std::string& destination,
+                const std::string& msg) {
+  struct sockaddr_un dst;
+  std::memset(&dst, 0, sizeof(dst));
+  if (destination.empty() || destination.size() >= sizeof(dst.sun_path))
+    return -EINVAL;
+  dst.sun_family = AF_UNIX;
+  std::strncpy(dst.sun_path, destination.data(), sizeof(dst.sun_path) - 1);
+  struct iovec iovec;
+  std::memset(&iovec, 0, sizeof(iovec));
+  iovec.iov_base = const_cast<char*>(msg.data());
+  iovec.iov_len = msg.size();
+  struct msghdr msghdr;
+  std::memset(&msghdr, 0, sizeof(msghdr));
+  msghdr.msg_name = &dst;
+  // For abstract domain sockets (i.e., where the first char is '@'), the
+  // terminating NUL byte is not counted towards the length, but any NUL
+  // bytes embedded in the path are. For example, in the extreme case it
+  // would be possible to have two different sockets named NUL and NUL NUL.
+  auto pathlen = destination[0] == '@' ? destination.size()
+                                       : std::strlen(&destination[0]) + 1;
+  msghdr.msg_namelen = offsetof(struct sockaddr_un, sun_path) + pathlen;
+  msghdr.msg_iov = &iovec;
+  msghdr.msg_iovlen = 1;
+  return ::sendmsg(socket, &msghdr, MSG_NOSIGNAL);
```

Comment:
Why do we need `MSG_NOSIGNAL` for sending a single packet?

Reply:
We use `MSG_NOSIGNAL` because we want to return a regular `caf::error` instead of getting `SIGPIPE` if the socket was shut down for writing. I don't think the number of packets is relevant for this.

Since `MSG_NOSIGNAL` seems to be linux-specific I've slightly refactored this to not hard-code `MSG_NOSIGNAL` but to instead take a generic `flags` argument.

Reply:
see: https://pubs.opengroup.org/onlinepubs/9699919799/functions/send.html:

> MSG_NOSIGNAL  
&nbsp;&nbsp;Requests not to send the SIGPIPE signal if an attempt to send is made on a **stream-oriented socket** that is no longer connected. The [EPIPE] error shall still be returned.

(emphasis mine).

Reply:
So the real question is if we need `MSG_NOSIGNAL` for a `SOCK_DGRAM` socket.

I didn't really find any sort of explanation by looking at the systemd git history, so this might just be an oversight or a workaround for some ancient kernel bug. Either way, I've removed it since the docs clearly say it should not be necessary.